db/cassandra/src/main/java/org/commonjava/indy/cassandra/data/CassandraStoreDataManager.java

```diff
@@ -173,6 +186,140 @@ protected ArtifactStore putArtifactStoreInternal( StoreKey storeKey, ArtifactSto
         return toArtifactStore( dtxArtifactStore );
     }
 
+    @Override
+    public Set<Group> affectedBy( Collection<StoreKey> keys ) throws IndyDataException
+    {
+
+        final Set<Group> result = new HashSet<>();
+
+        // use these to avoid recursion
+        final Set<StoreKey> processed = new HashSet<>();
+        final LinkedList<StoreKey> toProcess = new LinkedList<>( keys );
+
+        while ( !toProcess.isEmpty() )
+        {
+            StoreKey key = toProcess.removeFirst();
+            if ( key == null )
+            {
+                continue;
+            }
+
+            if ( processed.add( key ) )
+            {
+                DtxAffectedStore affectedStore = storeQuery.getAffectedStore( key );
+                Set<StoreKey> affected = affectedStore.getAffectedStoreKeys();
+                if ( affected != null )
+                {
+                    logger.debug( "Get affectedByStores, key: {}, affected: {}", key, affected );
+                    affected = affected.stream().filter( k -> k.getType() == group ).collect( Collectors.toSet() );
+                    for ( StoreKey gKey : affected )
+                    {
+                        // avoid loading the ArtifactStore instance again and again
+                        if ( !processed.contains( gKey ) && !toProcess.contains( gKey ) )
+                        {
+                            ArtifactStore store = getArtifactStoreInternal( gKey );
+
+                            // if this group is disabled, we don't want to keep loading it again and again.
+                            if ( store.isDisabled() )
+                            {
+                                processed.add( gKey );
+                            }
+                            else
+                            {
+                                // add the group to the toProcess list so we can find any result that might include it in their own membership
+                                toProcess.addLast( gKey );
+                                result.add( (Group) store );
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if ( logger.isTraceEnabled() )
+        {
+            logger.trace( "Groups affected by {} are: {}", keys,
+                          result.stream().map( ArtifactStore::getKey ).collect( Collectors.toSet() ) );
+        }
+        return filterAffectedGroups( result );
+    }
+
+    @Override
+    protected void refreshAffectedBy( final ArtifactStore store, final ArtifactStore original, StoreUpdateAction action )
+    {
+        if ( store == null )
+        {
+            return;
+        }
+
+        if ( store instanceof Group && isExcludedGroup( (Group) store ) )
+        {
+            logger.info( "Skip affectedBy calculation of group: {}", store.getName() );
+            return;
+        }
+
+        if ( action == DELETE )
```

Comment:
Seems this DELETE and STORE action code block has many duplicated lines with IninispanDataStoreManager, think about to extract them to one single method to avoid duplication?

Reply:
Good idea and done to move it into the AbstractStoreDataManager.